optimade/validator/validator.py

```diff
@@ -610,7 +627,12 @@ def _construct_single_property_filters(
             _test_value = self._format_test_value(test_value, prop_type, operator)
 
             query = f"{prop} {operator} {_test_value}"
-            response, _ = self._get_endpoint(f"{endp}?filter={query}", multistage=True)
+            response, _ = self._get_endpoint(
+                f"{endp}?filter={query}", multistage=True, optional=query_optional
+            )
+            if not response:
```

Comment:
`requests` will _always_ return a `Response` instance, right? It might not be jsonifiable, but it should still always return something, I guess?

Maybe this could be changed to a `try/except` for trying to jsonify the response? You should catch `json.JSONDecodeError` I believe.

Reply:
The problem here is that `_get_endpoint` is itself a "test case" and the response validation is handled inside `_get_endpoint` itself. In the case that the *actual* response doesn't have a valid return code, or can't be JSON'd, this method (as with all `@test_case` decorated methods) will return `(None, "string explaining what happened")`. So hopefully your suggestion is already handled further down the code.

Reply:
Ah. But if you're already testing if the response can be retrieved as a JSON, why not just pass this back? Returning the response code as well as an extra variable? This seems like we're doing a job twice, i.e., the jsonifying.

Reply:
Ah, I'm with you now. It does get passed back if it was successful, and I'm assuming that the JSONification would be cached at the `requests` level anyway (I'll check this). I don't think its worth overhauling that for this PR though.

Reply:
(This is all general to `@test_case` decorated functions, rather than _get_endpoint btw, which doesn't care about json - reading my initial response and it was a bit clumsy; most decorated methods cast to JSON [as this one does], but JSON decode errors are caught at the `@test_case` level.)

Reply:
Okay. It seems a bit convoluted, but I agree that it's not for this PR.

Reply:
Hopefully its just my description that is convoluted :grin: